File: src/android/app/src/main/java/org/citra/citra_emu/features/cheats/model/Cheat.java
package org.citra.citra_emu.features.cheats.model;

import androidx.annotation.Keep;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class Cheat {
    @Keep
    private final long mPointer;

    private Runnable mEnabledChangedCallback = null;

    @Keep
    private Cheat(long pointer) {
        mPointer = pointer;
    }

    @Override
    protected native void finalize();

    @NonNull
    public native String getName();

    @NonNull
    public native String getNotes();

    @NonNull
    public native String getCode();

    public native boolean getEnabled();

    public void setEnabled(boolean enabled) {
        setEnabledImpl(enabled);
        onEnabledChanged();
    }

    private native void setEnabledImpl(boolean enabled);

    public void setEnabledChangedCallback(@Nullable Runnable callback) {
        mEnabledChangedCallback = callback;
    }

    private void onEnabledChanged() {
        if (mEnabledChangedCallback != null) {
            mEnabledChangedCallback.run();
        }
    }

    /**
     * If the code is valid, returns 0. Otherwise, returns the 1-based index
     * for the line containing the error.
     */
    public static native int isValidGatewayCode(@NonNull String code);

    public static native Cheat createGatewayCode(@NonNull String name, @NonNull String notes,
                                                 @NonNull String code);
}

File: src/android/app/src/main/java/org/citra/citra_emu/features/cheats/model/Cheat.kt
// Copyright 2023 Citra Emulator Project
// Licensed under GPLv2 or any later version
// Refer to the license.txt file included.

package org.citra.citra_emu.features.cheats.model

import androidx.annotation.Keep

@Keep
class Cheat(@field:Keep private val mPointer: Long) {
    private var enabledChangedCallback: Runnable? = null
    protected external fun finalize()

    external fun getName(): String

    external fun getNotes(): String

    external fun getCode(): String

    external fun getEnabled(): Boolean

    fun setEnabled(enabled: Boolean) {
        setEnabledImpl(enabled)
        onEnabledChanged()
    }

    private external fun setEnabledImpl(enabled: Boolean)

    fun setEnabledChangedCallback(callback: Runnable) {
        enabledChangedCallback = callback
    }

    private fun onEnabledChanged() {
        enabledChangedCallback?.run()
    }

    companion object {
        /**
         * If the code is valid, returns 0. Otherwise, returns the 1-based index
         * for the line containing the error.
         */
        @JvmStatic
        external fun isValidGatewayCode(code: String): Int

        @JvmStatic
        external fun createGatewayCode(name: String, notes: String, code: String): Cheat
    }
}
